R/memoise.r

```diff
@@ -35,64 +35,132 @@
 #' # that the outcome is the same.
 #' memB(2, dummy="b")
-#' 
-#' # You can create multiple memoisations of the same function, 
-#' # and they'll be independent. 
+#'
+#' # You can create multiple memoisations of the same function,
+#' # and they'll be independent.
 #' memA(2)
 #' memA2 <- memoise(a)
 #' memA(2)  # Still the same outcome
 #' memA2(2) # Different cache, different outcome
-#' 
-#' # Don't do the same memoisation assignment twice: a brand-new 
-#' # memoised function also means a brand-new cache, and *that* 
+#'
+#' # Don't do the same memoisation assignment twice: a brand-new
+#' # memoised function also means a brand-new cache, and *that*
 #' # you could as easily and more legibly achieve using forget().
-#' # (If you're not sure whether you already memoised something,  
+#' # (If you're not sure whether you already memoised something,
 #' #  use is.memoised() to check.)
 #' memA(2)
 #' memA <- memoise(a)
 #' memA(2)
-memoise <- memoize <- function(f) {
+memoise <- memoize <- function(f, envir = parent.frame()) {
+  # We must not even try evaluating f -- once we start, there's no way back
+  if (inherits(try(eval.parent(substitute(f)), silent = TRUE), "try-error")) {
+    warning("Can't access f -- using old-style memoisation with dots interface. ",
+            "Define the memoised function before memoising to avoid this warning.")
+    memoise_old(f)
+  } else {
+    memoise_new(f, envir)
+  }
+}
+
+#' @importFrom stats setNames
+memoise_new <- function(f, envir) {
+  f_formals <- formals(f)
+  f_formal_names <- names(f_formals)
+  f_formal_name_list <- lapply(f_formal_names, as.name)
+
+  # list(...)
+  list_call <- as.call(c(list(as.name("list")), f_formal_name_list))
+
+  # memoised_function(...)
+  init_call_args <- setNames(f_formal_name_list, f_formal_names)
+  init_call <- as.call(c(as.name("memoised_function"), init_call_args))
+
   cache <- new_cache()
-  
+
+  memo_f <- eval(
+    bquote(function(...) {
+      hash <- digest(.(list_call))
+
+      if (cache$has_key(hash)) {
+        res <- cache$get(hash)
+      } else {
+        res <- withVisible(.(init_call))
+        cache$set(hash, res)
+      }
+
+      if (res$visible) {
+        res$value
+      } else {
+        invisible(res$value)
+      }
+    },
+    as.environment(list(list_call = list_call, init_call = init_call)))
+  )
+  formals(memo_f) <- f_formals
+  attr(memo_f, "memoised") <- TRUE
+
+  cache_env <- new.env(parent = envir)
+  cache_env$cache <- cache
+  cache_env$memoised_function <- f
+  cache_env$digest <- digest
+  environment(memo_f) <- cache_env
+
+  return(memo_f)
+}
+
+memoise_old <- function(f) {
+  cache <- new_cache()
+
   memo_f <- function(...) {
     hash <- digest(list(...))
-    
+
     if (cache$has_key(hash)) {
-      cache$get(hash)
+      res <- cache$get(hash)
     } else {
-      res <- f(...)
+      res <- withVisible(f(...))
       cache$set(hash, res)
-      res
+    }
+
+    if (res$visible) {
```

Comment:
Could you add a little reasoning about why you added this? To me, you're obviously only calling a memoised function for its return value (not its side effects) so visibility is not very important